### PR TITLE
[feat] pp 1/n

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Adds base RPC plugin for RPC based ddp communication.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -222,6 +219,20 @@ def __setstate__(self, d):
     def on_save(self, checkpoint):
         return checkpoint
 
+    @property
+    def rpc_enabled(self):
+        if self.ddp_plugin is not None and isinstance(self.ddp_plugin, RPCPlugin):
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator.py:224]:
```suggestion
       return self.ddp_plugin is not None and isinstance(self.ddp_plugin, RPCPlugin):
```

File: pytorch_lightning/plugins/rpc_plugin.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    def on_exit_rpc_process(self, trainer) -> None:
+        """
+        Called to exit RPC process that is being managed by main process.
+        Args:
+            trainer: The trainer object.
+        """
+        self.exit_rpc_process()
+
+    def exit_rpc_process(self):
```
Comment (tchaton) [inline pytorch_lightning/plugins/rpc_plugin.py:78]: Maybe can we moved to RPCPlugin.

File: pytorch_lightning/plugins/rpc_plugin.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    def on_main_rpc_connection(self, trainer) -> None:
+        """
+        Called when main rpc connection has been established.
+        Args:
+            trainer: The trainer object.
+        """
+        raise NotImplementedError
+
+    def on_exit_rpc_process(self, trainer) -> None:
```
Comment (tchaton) [inline pytorch_lightning/plugins/rpc_plugin.py:70]: Not sure why we have a `on_exit_rpc_process ` and `exit_rpc_process`.

File: pytorch_lightning/plugins/rpc_plugin.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        raise NotImplementedError
+
+    @property
+    def is_main_rpc_process(self) -> bool:
+        """
+        Override to add logic to determine current process is main RPC process.
+        """
+        raise NotImplementedError
+
+    def barrier(self) -> None:
```
Comment (tchaton) [inline pytorch_lightning/plugins/rpc_plugin.py:103]: barrier should take a name.

File: pytorch_lightning/plugins/rpc_plugin.py (hunk 3 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/rpc_plugin.py:71]: Needs to be separate so the user can call `exit_rpc_process` independently of the logic in pipe. Maybe the name should be `on_exit_accelerator_rpc_process` to be clearer

File: pytorch_lightning/plugins/rpc_plugin.py (hunk 2 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/rpc_plugin.py:79]: It is in the RPCPlugin

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/accelerators/accelerator.py [hunk 5]
```diff
@@ -146,10 +143,10 @@ def setup_optimizers(self, model):
         self.trainer.lr_schedulers = lr_schedulers
         self.trainer.optimizer_frequencies = optimizer_frequencies
 
-    def init_ddp_connection(
+    def init_distributed_connection(
```
Comment (williamFalcon) [inline pytorch_lightning/accelerators/accelerator.py:146]: is this backward compatible?

File: pytorch_lightning/accelerators/cpu_accelerator.py [hunk 6]
```diff
@@ -90,3 +90,7 @@ def sync_tensor(self,
                     group: Optional[Any] = None,
                     reduce_op: Optional[Union[ReduceOp, str]] = None) -> torch.Tensor:
         return tensor
+
+    @property
+    def is_ddp_based(self):
```
Comment (williamFalcon) [inline pytorch_lightning/accelerators/cpu_accelerator.py:95]: this seems very specific no?

**Review by williamFalcon — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py (hunk 5 above)
Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py]: It wasn't, reverted the change!

File: pytorch_lightning/accelerators/cpu_accelerator.py (hunk 6 above)
Comment (SeanNaren) [reply pytorch_lightning/accelerators/cpu_accelerator.py]: It was and the name wasn't very well defined. I've changed this to `require_distributed_sampler` which all accelerators now have control over, which makes it a lot cleaner when defining custom accelerators!

File: pytorch_lightning/accelerators/accelerator.py [hunk 7]
```diff
@@ -222,6 +219,18 @@ def __setstate__(self, d):
     def on_save(self, checkpoint):
         return checkpoint
 
+    @property
+    def rpc_enabled(self):
+        return self.ddp_plugin is not None and isinstance(self.ddp_plugin, RPCPlugin)
+
+    @property
+    def distributed_sampler_kwargs(self):
+        raise NotImplementedError
```
Comment (justusschock) [inline pytorch_lightning/accelerators/accelerator.py:228]: Do we want a default implementation for that? I.e. returning an empty duct here as default. I don't think having a property raising NotImplementedError in a property is good practice.

File: pytorch_lightning/accelerators/accelerator.py [hunk 8]
```diff
@@ -222,6 +219,18 @@ def __setstate__(self, d):
...
+
+    @property
+    def require_distributed_sampler(self):
+        raise NotImplementedError
```
Comment (justusschock) [inline pytorch_lightning/accelerators/accelerator.py:232]: Same as above. Can we make it return false per default?

**Review by justusschock — APPROVED** [review]
I like it. Only some really minor questions.

File: pytorch_lightning/accelerators/accelerator.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:232]: I think it is fine for both. An accelerator shouldn't be used. It is metaclass.
